File: src/paritygrid/application/planner/registry.py
"""Dependency-neutral contracts for the closed pipeline node registry."""

from __future__ import annotations

import re
from dataclasses import dataclass
from enum import StrEnum
from typing import cast

from paritygrid.application.ports.configuration import ConfigurationDocument
from paritygrid.domain.pipeline import NodeKind

NODE_REGISTRY_VERSION = 1
MAX_NODE_REGISTRY_DEFINITIONS = 128
MAX_NODE_CONFIGURATION_FIELDS = 64
MAX_NODE_CONFIGURATION_FIELD_LENGTH = 64

_CONFIGURATION_FIELD_PATTERN = re.compile(
    r"[a-z][a-z0-9]*(?:_[a-z0-9]+)*",
    flags=re.ASCII,
)


class NodeRegistryError(ValueError):
    """Base failure for registry definitions and lookups."""


class UnknownNodeKindError(NodeRegistryError):
    """A pipeline references a kind outside the closed registry."""


class UnsupportedNodeConfigurationVersionError(NodeRegistryError):
    """A node requests a configuration version the registry does not define."""


class InvalidNodeConfigurationError(NodeRegistryError):
    """A node configuration violates its registered structural schema."""


class NodeRole(StrEnum):
    """Closed semantic roles used by later planner validators."""

    SOURCE = "source"
    TRANSFORM = "transform"
    RECONCILIATION = "reconciliation"
    REPAIR_PLAN = "repair_plan"
    APPROVAL = "approval"
    REPAIR_EFFECT = "repair_effect"
    VERIFICATION = "verification"
    EXPORT = "export"


class ConnectorRequirement(StrEnum):
    """Connector relationship required by a node definition."""

    NONE = "none"
    SOURCE = "source"
    TARGET = "target"


class RetryBehavior(StrEnum):
    """Closed retry ownership declared by a node definition."""

    NEVER = "never"
    CONNECTOR = "connector"


class PlannerRunnerKind(StrEnum):
    """Runner families a compiled node may support."""

    SEQUENTIAL = "sequential"
    THREADED = "threaded"
    ASYNCIO = "asyncio"
    PROCESS = "process"


class NodeConfigurationValueKind(StrEnum):
    """Exact scalar kinds accepted by a registered configuration field."""

    BOOLEAN = "boolean"
    INTEGER = "integer"
    TEXT = "text"


@dataclass(frozen=True, slots=True, order=True)
class NodeConfigurationField:
    """One exact field in a versioned node configuration schema."""

    name: str
    value_kind: NodeConfigurationValueKind
    required: bool

    def __post_init__(self) -> None:
        name = cast(object, self.name)
        if not isinstance(name, str):
            raise TypeError("node configuration field name must be text")
        if not 1 <= len(name) <= MAX_NODE_CONFIGURATION_FIELD_LENGTH:
            raise NodeRegistryError("node configuration field name is outside the size limit")
        if not name.isascii() or _CONFIGURATION_FIELD_PATTERN.fullmatch(name) is None:
            raise NodeRegistryError("node configuration field name must use canonical snake case")
        _require_exact(
            self.value_kind,
            NodeConfigurationValueKind,
            "node configuration value kind",
        )
        if type(self.required) is not bool:
            raise TypeError("node configuration field required marker must be boolean")


@dataclass(frozen=True, slots=True, repr=False)
class NodeConfigurationSchema:
    """One immutable exact-key scalar configuration schema."""

    version: int
    fields: tuple[NodeConfigurationField, ...]

    def __post_init__(self) -> None:
        version = cast(object, self.version)
        if type(version) is not int:
            raise TypeError("node configuration schema version must be an integer")
        if not 1 <= version <= 2_147_483_647:
            raise NodeRegistryError(
                "node configuration schema version is outside the supported range"
            )
        fields = _require_exact_tuple(
            self.fields,
            NodeConfigurationField,
            "node configuration fields",
        )
        if len(fields) > MAX_NODE_CONFIGURATION_FIELDS:
            raise NodeRegistryError("node configuration schema exceeds the field limit")
        names = tuple(field.name for field in fields)
        _require_unique(names, "node configuration field names")
        object.__setattr__(self, "fields", tuple(sorted(fields, key=lambda field: field.name)))

    def validate(self, document: ConfigurationDocument) -> None:
        """Reject missing, unknown, or wrongly typed configuration fields."""
        _require_exact(document, ConfigurationDocument, "node configuration")
        mapping = document.to_mapping()
        definitions = {field.name: field for field in self.fields}
        missing = tuple(
            field.name for field in self.fields if field.required and field.name not in mapping
        )
        if missing:
            raise InvalidNodeConfigurationError("node configuration is missing required fields")
        if frozenset(mapping) - frozenset(definitions):
            raise InvalidNodeConfigurationError("node configuration contains unknown fields")
        if any(
            not _matches_value_kind(value, definitions[name].value_kind)
            for name, value in mapping.items()
        ):
            raise InvalidNodeConfigurationError(
                "node configuration contains an invalid field value"
            )

    def __repr__(self) -> str:
        return f"NodeConfigurationSchema(version={self.version!r}, fields={len(self.fields)})"


@dataclass(frozen=True, slots=True)
class NodeDefinition:
    """Complete dependency-neutral metadata for one closed node kind."""

    kind: NodeKind
    role: NodeRole
    configuration_schema: NodeConfigurationSchema
    connector_requirement: ConnectorRequirement
    supported_runners: tuple[PlannerRunnerKind, ...]
    retry_behavior: RetryBehavior
    requires_idempotency: bool

    def __post_init__(self) -> None:
        _require_exact(self.kind, NodeKind, "node definition kind")
        _require_exact(self.role, NodeRole, "node definition role")
        _require_exact(
            self.configuration_schema,
            NodeConfigurationSchema,
            "node definition configuration schema",
        )
        _require_exact(
            self.connector_requirement,
            ConnectorRequirement,
            "node definition connector requirement",
        )
        runners = _require_exact_tuple(
            self.supported_runners,
            PlannerRunnerKind,
            "node definition supported runners",
        )
        if not runners:
            raise NodeRegistryError("node definition requires at least one supported runner")
        _require_unique(runners, "node definition supported runners")
        _require_exact(self.retry_behavior, RetryBehavior, "node definition retry behavior")
        if type(self.requires_idempotency) is not bool:
            raise TypeError("node definition idempotency marker must be boolean")
        if self.role is NodeRole.REPAIR_EFFECT and not self.requires_idempotency:
            raise NodeRegistryError("repair effect nodes must require idempotency")
        object.__setattr__(
            self,
            "supported_runners",
            tuple(sorted(runners, key=lambda runner: runner.value)),
        )


@dataclass(frozen=True, slots=True, repr=False)
class NodeRegistry:
    """An immutable versioned set of unique node definitions."""

    definitions: tuple[NodeDefinition, ...]
    version: int = NODE_REGISTRY_VERSION

    def __post_init__(self) -> None:
        version = cast(object, self.version)
        if type(version) is not int:
            raise TypeError("node registry version must be an integer")
        if version != NODE_REGISTRY_VERSION:
            raise NodeRegistryError("node registry version is unsupported")
        definitions = _require_exact_tuple(
            self.definitions,
            NodeDefinition,
            "node registry definitions",
        )
        if not definitions:
            raise NodeRegistryError("node registry requires at least one definition")
        if len(definitions) > MAX_NODE_REGISTRY_DEFINITIONS:
            raise NodeRegistryError("node registry exceeds the definition limit")
        kinds = tuple(definition.kind for definition in definitions)
        _require_unique(kinds, "node registry kinds")
        object.__setattr__(
            self,
            "definitions",
            tuple(sorted(definitions, key=lambda definition: str(definition.kind))),
        )

    def get(self, kind: NodeKind) -> NodeDefinition | None:
        """Return one exact kind definition, if registered."""
        _require_exact(kind, NodeKind, "node kind")
        return next(
            (definition for definition in self.definitions if definition.kind == kind), None
        )

    def require(self, kind: NodeKind, configuration_version: int) -> NodeDefinition:
        """Resolve one kind and exact registered configuration version."""
        definition = self.get(kind)
        if definition is None:
            raise UnknownNodeKindError("pipeline node kind is not registered")
        version = cast(object, configuration_version)
        if type(version) is not int:
            raise TypeError("node configuration version must be an integer")
        if version != definition.configuration_schema.version:
            raise UnsupportedNodeConfigurationVersionError(
                "pipeline node configuration version is unsupported"
            )
        return definition

    def __repr__(self) -> str:
        return f"NodeRegistry(version={self.version!r}, definitions={len(self.definitions)})"


def _matches_value_kind(value: object, kind: NodeConfigurationValueKind) -> bool:
    if kind is NodeConfigurationValueKind.BOOLEAN:
        return type(value) is bool
    if kind is NodeConfigurationValueKind.INTEGER:
        return type(value) is int
    return isinstance(value, str)


def _require_exact(value: object, expected: type[object], subject: str) -> None:
    if type(value) is not expected:
        raise TypeError(f"{subject} must use {expected.__name__}")


def _require_exact_tuple[T](value: object, item_type: type[T], subject: str) -> tuple[T, ...]:
    if not isinstance(value, tuple):
        raise TypeError(f"{subject} must be a tuple")
    items = cast(tuple[object, ...], value)
    if any(type(item) is not item_type for item in items):
        raise TypeError(f"{subject} contains an invalid value")
    return cast(tuple[T, ...], items)


def _require_unique(values: tuple[object, ...], subject: str) -> None:
    if len(set(values)) != len(values):
        raise NodeRegistryError(f"{subject} must be unique")

File: web/src/pipelines/registry.ts
/**
 * Closed, exhaustive frontend mirror of the accepted backend node registry
 * (`src/paritygrid/application/planner/registry.py`, registry version 1).
 *
 * Every node kind the backend planner accepts is defined here exactly once
 * with its configuration schema, typed ports, connector requirement,
 * supported runners, and capability requirements. Nothing outside this
 * registry can be added to a draft, rendered on the canvas, inspected, or
 * published; unknown kinds are rejected at parse time rather than preserved
 * as editable "unknown" objects.
 *
 * When the backend registry changes, this file must change in the same
 * change-set; the exhaustive tests assert the counts and key invariants.
 */

export const NODE_REGISTRY_VERSION = 1;

/** Exact closed logical payload types exchanged between built-in nodes. */
export const PortValueType = {
  RawRecords: "records.raw",
  NormalizedRecords: "records.normalized",
  ValidatedRecords: "records.validated",
  PartitionedRecords: "records.partitioned",
  Reconciliation: "reconciliation.result",
  RepairPlan: "repair.plan",
  ApprovedRepairPlan: "repair.plan.approved",
  RepairResult: "repair.result",
  Verification: "verification.result",
} as const;
export type PortValueType = (typeof PortValueType)[keyof typeof PortValueType];

export type NodeRole =
  | "source"
  | "transform"
  | "reconciliation"
  | "repair_plan"
  | "approval"
  | "repair_effect"
  | "verification"
  | "export";

export type ConnectorRequirement = "none" | "source" | "target";

export type RunnerKind = "sequential" | "threaded" | "asyncio" | "process";

export type RetryBehavior = "never" | "connector";

/** Closed connector behaviors, mirroring the backend capability set. */
export const ConnectorCapability = {
  Read: "read",
  Write: "write",
  AsyncIo: "async_io",
  BlockingIo: "blocking_io",
  Idempotency: "idempotency",
  SchemaDiscovery: "schema_discovery",
} as const;
export type ConnectorCapability =
  (typeof ConnectorCapability)[keyof typeof ConnectorCapability];

export interface InputPortDefinition {
  readonly name: string;
  readonly accepts: readonly PortValueType[];
  readonly required: boolean;
  readonly maximumConnections: number;
}

export interface OutputPortDefinition {
  readonly name: string;
  readonly valueType: PortValueType;
}

export interface PortSchema {
  readonly inputs: readonly InputPortDefinition[];
  readonly outputs: readonly OutputPortDefinition[];
}

/** One exact scalar field in a versioned node configuration schema. */
export interface ConfigurationField {
  readonly name: string;
  readonly valueKind: "boolean" | "integer" | "text";
  readonly required: boolean;
  readonly minimum?: number;
  readonly maximum?: number;
}

export interface ConfigurationSchema {
  readonly version: number;
  readonly fields: readonly ConfigurationField[];
}

export interface NodeDefinition {
  readonly kind: string;
  readonly role: NodeRole;
  readonly configurationSchema: ConfigurationSchema;
  readonly ports: PortSchema;
  readonly connectorRequirement: ConnectorRequirement;
  /** Exact capability keys a bound connector must support. */
  readonly requiredCapabilities: readonly ConnectorCapability[];
  readonly supportedRunners: readonly RunnerKind[];
  readonly retryBehavior: RetryBehavior;
  readonly requiresIdempotency: boolean;
  /** Short human explanation used by the canvas and inspector. */
  readonly label: string;
  readonly description: string;
}

const recordsInput = (accepts: readonly PortValueType[]): InputPortDefinition[] => [
  { name: "records", accepts, required: true, maximumConnections: 1 },
];

const singleOutput = (
  name: string,
  valueType: PortValueType,
): OutputPortDefinition[] => [{ name, valueType }];

const field = (
  name: string,
  valueKind: ConfigurationField["valueKind"],
  limits: { minimum?: number; maximum?: number } = {},
): ConfigurationField => ({ name, valueKind, required: false, ...limits });

const configuration = (
  version: number,
  fields: readonly ConfigurationField[],
): ConfigurationSchema => ({
  version,
  fields,
});

const emptyConfigurationV1 = configuration(1, []);

const httpSourceConfigurationV1 = configuration(1, [
  field("page_size", "integer", { minimum: 1, maximum: 10_000 }),
]);

const csvSourceConfigurationV1 = configuration(1, [
  field("encoding", "text"),
  field("header", "boolean"),
]);

const jsonlSourceConfigurationV1 = configuration(1, [field("encoding", "text")]);

const partitionConfigurationV1 = configuration(1, [
  field("partition_count", "integer", { minimum: 1, maximum: 10_000 }),
]);

const exportConfigurationV1 = configuration(1, [field("compression", "text")]);

const standardRunners: readonly RunnerKind[] = ["asyncio", "sequential", "threaded"];
const cpuRunners: readonly RunnerKind[] = [
  "asyncio",
  "process",
  "sequential",
  "threaded",
];

/**
 * The exhaustive definitions. Sorted by kind for deterministic iteration;
 * the shapes mirror the backend `_definition(...)` call sites exactly.
 */
export const NODE_DEFINITIONS: readonly NodeDefinition[] = [
  {
    kind: "export.parquet",
    role: "export",
    configurationSchema: exportConfigurationV1,
    ports: {
      inputs: recordsInput([
        PortValueType.RawRecords,
        PortValueType.NormalizedRecords,
        PortValueType.ValidatedRecords,
        PortValueType.PartitionedRecords,
      ]),
      outputs: [],
    },
    connectorRequirement: "none",
    requiredCapabilities: [],
    supportedRunners: standardRunners,
    retryBehavior: "never",
    requiresIdempotency: false,
    label: "Export Parquet",
    description: "Writes verified records to a content-addressed Parquet artifact.",
  },
  {
    kind: "reconcile.target",
    role: "reconciliation",
    configurationSchema: emptyConfigurationV1,
    ports: {
      inputs: recordsInput([
        PortValueType.NormalizedRecords,
        PortValueType.ValidatedRecords,
        PortValueType.PartitionedRecords,
      ]),
      outputs: singleOutput("reconciliation", PortValueType.Reconciliation),
    },
    connectorRequirement: "target",
    requiredCapabilities: [ConnectorCapability.Read],
    supportedRunners: standardRunners,
    retryBehavior: "connector",
    requiresIdempotency: false,
    label: "Reconcile with target",
    description: "Compares pipeline records against the bound target connector.",
  },
  {
    kind: "repair.apply",
    role: "repair_effect",
    configurationSchema: emptyConfigurationV1,
    ports: {
      inputs: [
        {
          name: "approved-plan",
          accepts: [PortValueType.ApprovedRepairPlan],
          required: true,
          maximumConnections: 1,
        },
      ],
      outputs: singleOutput("repair-result", PortValueType.RepairResult),
    },
    connectorRequirement: "target",
    requiredCapabilities: [ConnectorCapability.Idempotency, ConnectorCapability.Write],
    supportedRunners: standardRunners,
    retryBehavior: "connector",
    requiresIdempotency: true,
    label: "Apply repairs",
    description: "Applies the approved repair plan idempotently.",
  },
  {
    kind: "repair.approval",
    role: "approval",
    configurationSchema: emptyConfigurationV1,
    ports: {
      inputs: [
        {
          name: "repair-plan",
          accepts: [PortValueType.RepairPlan],
          required: true,
          maximumConnections: 1,
        },
      ],
      outputs: singleOutput("approved-plan", PortValueType.ApprovedRepairPlan),
    },
    connectorRequirement: "none",
    requiredCapabilities: [],
    supportedRunners: standardRunners,
    retryBehavior: "never",
    requiresIdempotency: false,
    label: "Approval gate",
    description: "Explicit human approval before any repair effect.",
  },
  {
    kind: "repair.generate",
    role: "repair_plan",
    configurationSchema: emptyConfigurationV1,
    ports: {
      inputs: [
        {
          name: "reconciliation",
          accepts: [PortValueType.Reconciliation],
          required: true,
          maximumConnections: 1,
        },
      ],
      outputs: singleOutput("repair-plan", PortValueType.RepairPlan),
    },
    connectorRequirement: "none",
    requiredCapabilities: [],
    supportedRunners: cpuRunners,
    retryBehavior: "never",
    requiresIdempotency: false,
    label: "Generate repair plan",
    description: "Plans bounded non-destructive repairs from the reconciliation.",
  },
  {
    kind: "source.csv",
    role: "source",
    configurationSchema: csvSourceConfigurationV1,
    ports: { inputs: [], outputs: singleOutput("records", PortValueType.RawRecords) },
    connectorRequirement: "source",
    requiredCapabilities: [ConnectorCapability.Read],
    supportedRunners: standardRunners,
    retryBehavior: "connector",
    requiresIdempotency: false,
    label: "CSV source",
    description: "Reads CSV files through a bound source connector.",
  },
  {
    kind: "source.http.async",
    role: "source",
    configurationSchema: httpSourceConfigurationV1,
    ports: { inputs: [], outputs: singleOutput("records", PortValueType.RawRecords) },
    connectorRequirement: "source",
    requiredCapabilities: [ConnectorCapability.AsyncIo, ConnectorCapability.Read],
    supportedRunners: standardRunners,
    retryBehavior: "connector",
    requiresIdempotency: false,
    label: "Async HTTP source",
    description: "Reads a paginated asynchronous HTTP API through a connector.",
  },
  {
    kind: "source.http.blocking",
    role: "source",
    configurationSchema: httpSourceConfigurationV1,
    ports: { inputs: [], outputs: singleOutput("records", PortValueType.RawRecords) },
    connectorRequirement: "source",
    requiredCapabilities: [ConnectorCapability.BlockingIo, ConnectorCapability.Read],
    supportedRunners: standardRunners,
    retryBehavior: "connector",
    requiresIdempotency: false,
    label: "Blocking HTTP source",
    description: "Reads a blocking HTTP API through a bound connector.",
  },
  {
    kind: "source.jsonl",
    role: "source",
    configurationSchema: jsonlSourceConfigurationV1,
    ports: { inputs: [], outputs: singleOutput("records", PortValueType.RawRecords) },
    connectorRequirement: "source",
    requiredCapabilities: [ConnectorCapability.Read],
    supportedRunners: standardRunners,
    retryBehavior: "connector",
    requiresIdempotency: false,
    label: "JSON Lines source",
    description: "Reads JSON Lines files through a bound source connector.",
  },
  {
    kind: "transform.normalize",
    role: "transform",
    configurationSchema: emptyConfigurationV1,
    ports: {
      inputs: recordsInput([PortValueType.RawRecords]),
      outputs: singleOutput("records", PortValueType.NormalizedRecords),
    },
    connectorRequirement: "none",
    requiredCapabilities: [],
    supportedRunners: cpuRunners,
    retryBehavior: "never",
    requiresIdempotency: false,
    label: "Normalize records",
    description: "Normalizes raw records into the canonical record shape.",
  },
  {
    kind: "transform.partition",
    role: "transform",
    configurationSchema: partitionConfigurationV1,
    ports: {
      inputs: recordsInput([PortValueType.ValidatedRecords]),
      outputs: singleOutput("records", PortValueType.PartitionedRecords),
    },
    connectorRequirement: "none",
    requiredCapabilities: [],
    supportedRunners: cpuRunners,
    retryBehavior: "never",
    requiresIdempotency: false,
    label: "Partition records",
    description: "Partitions validated records into bounded work units.",
  },
  {
    kind: "transform.validate",
    role: "transform",
    configurationSchema: emptyConfigurationV1,
    ports: {
      inputs: recordsInput([PortValueType.NormalizedRecords]),
      outputs: singleOutput("records", PortValueType.ValidatedRecords),
    },
    connectorRequirement: "none",
    requiredCapabilities: [],
    supportedRunners: cpuRunners,
    retryBehavior: "never",
    requiresIdempotency: false,
    label: "Validate records",
    description: "Quarantines invalid records without losing valid work.",
  },
  {
    kind: "verify.target",
    role: "verification",
    configurationSchema: emptyConfigurationV1,
    ports: {
      inputs: [
        {
          name: "repair-result",
          accepts: [PortValueType.RepairResult],
          required: true,
          maximumConnections: 1,
        },
      ],
      outputs: singleOutput("verification", PortValueType.Verification),
    },
    connectorRequirement: "target",
    requiredCapabilities: [ConnectorCapability.Read],
    supportedRunners: standardRunners,
    retryBehavior: "connector",
    requiresIdempotency: false,
    label: "Verify target state",
    description: "Proves the target reached the expected logical state.",
  },
];

const BY_KIND = new Map(
  NODE_DEFINITIONS.map((definition) => [definition.kind, definition]),
);

/** Resolve one definition, or undefined for kinds outside the registry. */
export function nodeDefinition(kind: string): NodeDefinition | undefined {
  return BY_KIND.get(kind);
}

/** True when the port type of `source` may feed `target`. */
export function portsAreCompatible(
  source: OutputPortDefinition,
  target: InputPortDefinition,
): boolean {
  return target.accepts.includes(source.valueType);
}

/** Closed connector capability requirement for one registered kind. */
export function requiredConnectorCapabilities(
  kind: string,
): readonly ConnectorCapability[] {
  return nodeDefinition(kind)?.requiredCapabilities ?? [];
}
